### ISSUE TITLE
(Self-Voicing) Read Me Buttons Accordion Box

### ISSUE THREAD

**Issue by terracoda** [phetsims/gravity-force-lab-basics#219]
As discussed in today's Self-Voicing Research and Design Meeting.

![Screen Shot 2020-07-29 at 6 47 27 PM](https://user-images.githubusercontent.com/1902116/88854801-6bc7a100-d1cc-11ea-84d7-daadc419f221.png)

Please replace speech bubble with an X, like in the Mute Sound Button when "Stop Speech" button is pressed.

**Comment by terracoda** [comment]
@jessegreenberg, I'll sort out the accessible version asap.

**Comment by jessegreenberg** [comment]
Nice! @terracoda do you have a preference for where this should be placed in the sim? How exact-to-mockup should this be, is it OK if I use default PhET RectangularPushButtons instead of blue rectangle style in this image?

**Comment by terracoda** [comment]
@jessegreenberg, great questions. I'll get today:
- Yes on RectangularPushButtons
- making it draggable might complicate the interaction, depending on type of input, so maybe let's not do that.
- will it fit to the left of the Reset All button? And maybe open up rather than down?
- can you make the box go over (obscure) other content like the PhET Menu does? I don't think that is the normal way an accordionbox works, but can you make it do that?

I haven't given you the content in writing, and I can do that today.

**Comment by jessegreenberg** [comment]
> Please replace speech bubble with an X, like in the Mute Sound Button when "Stop Speech" button is pressed.

What is the behavior of this button, will it stop all speech  until pressed again or will it just silence the current utterance that is being spoken?

**Comment by jessegreenberg** [comment]
> And maybe open up rather than down?

It looks like ComboBox doesn't support this, but what is shown here is more of a Panel that becomes visible with an ExpandCollapse button, especially since it has no title. I can do something more custom for now that doesn't use AccordionBox.

> an you make the box go over (obscure) other content like the PhET Menu does?

Yes, it can obscure other content.

**Comment by jessegreenberg** [comment]
@terracoda are there assets for the T and speech bubble icon I can use? Where did those come from?

**Comment by jessegreenberg** [comment]
OK, boilerplate has been added for this. We have a temp icon, panel with buttons, control over enabling/disabling speech, visibility changes with self-voicing selection. Ready for design of button behavior or other assets/changes.

**Comment by terracoda** [comment]
This is really cool @jessegreenberg.

Here are some images for the actual icon
![self-voicing-BW](https://user-images.githubusercontent.com/1902116/89580214-94c0e500-d80f-11ea-8c7a-ca980f507d1e.png)
![self-voicing-GS](https://user-images.githubusercontent.com/1902116/89580220-97233f00-d80f-11ea-93e4-fa7415765a83.png)
![self-voicing-black](https://user-images.githubusercontent.com/1902116/89580264-a1453d80-d80f-11ea-928e-0977a63e831e.png)

When you are able to download the image, please change to either the grey or black icon.

**Comment by terracoda** [comment]
As for content for the buttons, here's what we can do:
- Hint Please! - Voice: "Change position and mass of spheres to explore."

- Sim Overview - Voice the static part of the screen summary. This may be challenging because of the separate paragraphs: 
    - "Gravity Force Lab: Basics is an interactive sim. It changes as you play with it. It has a Play Area and a Control Area. 
The Play Area has two mass spheres, a blue sphere labeled mass 1 and a red sphere labeled mass 2. A force arrow starts at the center of each sphere and points directly at the opposite sphere. 
Spheres can be moved closer or farther from each other along a track. The mass of each sphere can be increased or decreased. Each sphere is held in place by a robot. 
In the Control Area, checkboxes change what things are shown and how they behave, and a button resets the sim." 

- Current Details - Voice the dynamic part of the screen summary. This may be challenging because of the bullets.
    - "Currently, force on mass 1 by mass 2 is of equal size and pointing directly opposite to the force on mass 2 by mass Force arrows are tiny at 33.4 newtons.
mass 1 and mass 2 are close to each other, centers exactly 4 kilometers apart.
Mass of mass 1 is 2 billion kilograms, half the size of mass 2 at 4 billion kilograms.
Robots pull very little keeping spheres in place."

I have a feeling, that when the Overview or Current details buttons are pressed, the content shows up in a text box on the side of the sim and the text highlights as it is read.

If you could just implement the reading of the content that would be great. We can discuss boxes a text at our next design meeting.

**Comment by terracoda** [comment]
@jessegreenberg, the read me buttons sound good. One problem is that the Sim Overview, m1 and m2 are mentioned instead of mass 1 and mass 2.

This used to happen a lot in this sim because of the shared code between all the GFL regular and all the ISLC sims.

If you could check how the Sim Overview is pulling in strings, you might find you need to do something special to get "mass 1" and "mass 2" in the sentence that starts with, "The Play has two mass spheres, a blue sphere labeled m1 and a red sphere labeled m2."

If it is not obvious, you might need to reach out to @zepumph.

Otherwise, the buttons sound great.

And, yes, please change the Stop Speech button to a regular button that stops the current speech.

If easy, add a toggle button with the label Self-Voicing to toggle the Self-Voicing on and off. When off, could you add an X over the speech bubble. Don't fuss with the visuals too much because they will change. We just need a simple indication that the feature is off.

**Comment by terracoda** [comment]
I'll leave navigation of the buttons to a seperate issue.

**Comment by jessegreenberg** [comment]
> One problem is that the Sim Overview, m1 and m2 are mentioned instead of mass 1 and mass 2.

I see, I was using GFL strings instead of GFLB strings for this content. Should be fixed with the above commits.

**Comment by jessegreenberg** [comment]
> please change the Stop Speech button to a regular button that stops the current speech.

> add a toggle button with the label Self-Voicing to toggle the Self-Voicing on and off. When off, could you add an X over the speech bubble.

Both done in the above commits. I think this is ready or review @terracoda.

**Comment by jessegreenberg** [comment]
@terracoda I also made it so that wen the speaker is disabled, the other buttons in the panel become disabled and the speaker highlights do not appear.

**Comment by terracoda** [comment]
@jessegreenberg, this is working great.
When we had our "menu meeting", the team talked about have this kind of sub menu on the left edge of the sim.

Can you move it over to the left of the screen?
Also, wondering if the buttons can slide out horizontally, rather than vertically.

Kind of like this - ignore label changes:
![sketch-submenu](https://user-images.githubusercontent.com/1902116/90675134-a43f2580-e234-11ea-86ad-072b73eec663.jpg)

**Comment by terracoda** [comment]
Keep it at the same level vertically, but move it to the far left of the sim.

**Comment by terracoda** [comment]
@jessegreenberg, are the pictures above clear enough to move the menu over to the left "bounds"?

**Comment by terracoda** [comment]
![Screen Shot 2020-09-02 at 7 12 03 PM](https://user-images.githubusercontent.com/1902116/92042264-696cdf80-ed54-11ea-8b8e-f331e2e0aaac.png)
<img width="333" alt="Screen Shot 2020-09-02 at 7 39 44 PM" src="https://user-images.githubusercontent.com/1902116/92042272-6d98fd00-ed54-11ea-8c2d-c52ec843f1ee.png">

Maybe put the green plus in lower right corner.

**Comment by terracoda** [comment]
Very rough sketch in the previous comment.
Maybe the pop-out can be the same height as the icon. Not sure if that will be tall enough to accommodate the buttons.

But don't worry too much about it as this is just a prototype.

**Comment by terracoda** [comment]
Self-Voicing Content for this menu, assuming this is the pattern of an Accordion Box, we could try:

Self-Voicing Quick Menu (collapsed)
Self-Voicing Quick Menu (expanded)
Hint Please!
Sim Overview
Current Details
Stop Speech
Mute Speech

Use aria-expanded attribute to communicate the state. Ratio and  Proportion should have an implemented example.

**Comment by terracoda** [comment]
Over to you @jessegreenberg.
If these additional changes can't be done for Monday's interview, not to worry. Just get to it when you can.

**Comment by terracoda** [comment]
No aria in self-voicing...so let me re-think previous comment

Show Self-Voicing Quick Menu (when collapsed)
Hide Self-Voicing Quick Menu (when expanded)
Hint Please!
Sim Overview
Current Details
Stop Speech
Mute Speech

**Comment by jessegreenberg** [comment]
It seems like that would work pretty well for focus (read before interaction) but may not work for mouse.  For instance, user would click the button to open the panel and it would say "Hide Self-Voicing Quick Menu". Is that intended?

**Comment by jessegreenberg** [comment]
Ah, maybe it was assumed that all of those would only be read on focus. Because the behavior of the button would be to speak its hint/details/overview content, not its name.

**Comment by jessegreenberg** [comment]
OK, the above commits adds this. The button names are spoken. The behavior for the expand/collapse button is

- On Focus:
  - "Show Self-Voicing Quick Menu" (when closed)
  - "Hide Self-Voicing Quick Menu" (when open)

- On Button Press:
  - "Self Voicing Quick Menu hidden" (when closed)
  - "Self Voicing Quick Menu shown" (when open)

Layout and focus order was changed so that the ExpandCollapseButton comes first.

**Comment by terracoda** [comment]
@jessegreenberg, could you swap the order of the buttons so Current Details comes after Sim Overview?

**Comment by jessegreenberg** [comment]
The order of these buttons has been changed.

@terracoda should this issue stay open or do you think this can be closed now? For further requests we could open new issues.

**Comment by terracoda** [comment]
Verified. Thanks!
